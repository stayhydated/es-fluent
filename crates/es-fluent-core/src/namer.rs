use derive_more::{Debug, Deref, Display};
use heck::ToSnakeCase as _;
use quote::format_ident;

#[derive(Clone, Debug, Deref, Display, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct FluentKey(pub String);

impl FluentKey {
    pub const DELIMITER: &str = "-";
    pub fn new(ftl_name: &syn::Ident, sub_name: &str) -> Self {
        let normalized_name = ftl_name.to_string().to_snake_case();
        if sub_name.is_empty() {
            FluentKey(normalized_name)
        } else {
            FluentKey(format!(
                "{}{}{}",
                normalized_name,
                Self::DELIMITER,
                sub_name
            ))
        }
    }
}

impl quote::ToTokens for FluentKey {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let key_string = &self.0;
        tokens.extend(quote::quote! { #key_string });
    }
}

#[derive(Clone, Debug, Deref, Display, Eq, Hash, PartialEq)]
pub struct FluentDoc(pub String);

impl From<&FluentKey> for FluentDoc {
    fn from(ftl_key: &FluentKey) -> Self {
        FluentDoc(format!("Key = `{}`", *ftl_key))
    }
}

impl quote::ToTokens for FluentDoc {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let doc_string = &self.0;
        tokens.extend(quote::quote! { #doc_string });
    }
}

pub struct UnnamedItem(usize);

impl std::fmt::Display for UnnamedItem {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", Self::UNNAMED_PREFIX, self.0)
    }
}

impl UnnamedItem {
    const UNNAMED_PREFIX: &str = "f";

    pub fn to_ident(&self) -> syn::Ident {
        format_ident!("{}", self.to_string())
    }
}

impl From<usize> for UnnamedItem {
    fn from(index: usize) -> Self {
        Self(index)
    }
}
