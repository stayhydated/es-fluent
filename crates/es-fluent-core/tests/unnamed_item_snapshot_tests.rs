use es_fluent_core::namer::UnnamedItem;

#[test]
fn unnamed_item_formats_snapshot() {
    let indices = [
        0usize, 1, 2, 3, 4, 5, 7, 8, 9, 10, 15, 31, 32, 63, 64, 99, 100, 101, 255, 256, 511, 512,
        1023, 1024,
    ];

    let cases: Vec<(usize, String, String)> = indices
        .iter()
        .map(|&i| {
            let item: UnnamedItem = i.into();
            let display = item.to_string();
            let ident_str = item.to_ident().to_string();
            (i, display, ident_str)
        })
        .collect();

    insta::assert_ron_snapshot!("unnamed_item_formats_snapshot", &cases);
}
