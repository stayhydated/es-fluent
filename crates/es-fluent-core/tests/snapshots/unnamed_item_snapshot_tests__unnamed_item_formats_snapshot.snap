---
source: crates/es-fluent-core/tests/unnamed_item_snapshot_tests.rs
expression: "&cases"
---
[
  (0, "f0", "f0"),
  (1, "f1", "f1"),
  (2, "f2", "f2"),
  (3, "f3", "f3"),
  (4, "f4", "f4"),
  (5, "f5", "f5"),
  (7, "f7", "f7"),
  (8, "f8", "f8"),
  (9, "f9", "f9"),
  (10, "f10", "f10"),
  (15, "f15", "f15"),
  (31, "f31", "f31"),
  (32, "f32", "f32"),
  (63, "f63", "f63"),
  (64, "f64", "f64"),
  (99, "f99", "f99"),
  (100, "f100", "f100"),
  (101, "f101", "f101"),
  (255, "f255", "f255"),
  (256, "f256", "f256"),
  (511, "f511", "f511"),
  (512, "f512", "f512"),
  (1023, "f1023", "f1023"),
  (1024, "f1024", "f1024"),
]
